arangod/IResearch/Containers.h

```diff
@@ -49,70 +51,96 @@ struct typelist;
+    T const* operator->() const noexcept { return get(); }
 
-    explicit operator bool() const noexcept { return nullptr != get(); }
+    explicit operator bool() const noexcept { return get() != nullptr; }
 
-    bool ownsLock() const noexcept { return _lock.owns_lock(); }
+    ~Value() {
+      if (_self) {
+        _self->destroy();
+      }
+    }
 
    private:
-    friend class AsyncValue<T>;
+    friend class AsyncValue;
+    constexpr Value(AsyncValue& self) noexcept : _self{&self} {}
 
-    Value(std::shared_lock<std::shared_mutex>&& lock, T* resource)
-        : _lock{std::move(lock)}, _resource{resource} {}
-
-    std::shared_lock<std::shared_mutex> _lock;
-    T* _resource{};
+    AsyncValue* _self;
   };
 
-  explicit AsyncValue(T* resource) noexcept : _resource{resource} {}
+  explicit AsyncValue(T* resource) noexcept
+      : _resource{resource}, _count{[&] {
+          if (!resource) {
+            return kReset | kDestroy;
+          } else {
+            return kRef;
+          }
+        }()} {}
 
   ~AsyncValue() { reset(); }
 
-  auto lock() const {
-    auto lock = irs::make_shared_lock(_mutex);
-    return Value{std::move(lock), _resource};
+  [[nodiscard]] bool empty() const noexcept {
+    return _count.load(std::memory_order_acquire) & kReset;
   }
 
-  auto try_lock() const {
-    auto lock = irs::make_shared_lock(_mutex, std::try_to_lock);
-
-    if (lock.owns_lock()) {
-      return Value{std::move(lock), _resource};
+  [[nodiscard]] Value lock() noexcept {
+    if (empty()) {
+      return {};
     }
-
-    return Value{};
+    if (_count.fetch_add(kRef, std::memory_order_acquire) & kDestroy) {
+      return {};
+    }
+    return {*this};
   }
 
-  // will block until a write lock can be acquired on the _mutex
   void reset() {
-    auto lock = irs::make_unique_lock(_mutex);
-    _resource = nullptr;
+    auto count = _count.fetch_or(kReset, std::memory_order_acq_rel);
+    if (!(count & kReset)) {
+      destroy();
+    }
+    std::unique_lock lock{_m};
+    while (_resource != nullptr) {
+      _cv.wait(lock);
+    }
   }
 
-  bool empty() const {
-    auto lock = irs::make_shared_lock(_mutex);
-    return nullptr == _resource;
+ private:
+  static constexpr uint64_t kReset = 1;
+  static constexpr uint64_t kDestroy = 2;
+  static constexpr uint64_t kRef = 4;
+
+  void destroy() {
+    auto count = _count.fetch_sub(kRef, std::memory_order_acq_rel) - kRef;
+    if (count == kReset &&
+        _count.compare_exchange_strong(count, kReset | kDestroy,
+                                       std::memory_order_release,
+                                       std::memory_order_relaxed)) {
+      std::lock_guard lock{_m};
+      _resource = nullptr;
+      _cv.notify_all();  // should be under lock
+    }
   }
 
- private:
-  mutable std::shared_mutex _mutex;  // read-lock to prevent '_resource' reset()
   T* _resource;
+  mutable std::atomic_uint64_t _count;
+  std::mutex _m;
+  std::condition_variable _cv;
```

Comment:
In theory we can use _count as mutex and condvar, but it's not easy so I think it's ok